Add IIS Tool contrib package

bknueven opened the pull request:
## Fixes # N/A

## Summary/Motivation:
This formalizes a workflow I've used for a while to compute the IIS of an infeasible Pyomo model using the persistent solver interface of a commercial solver.

## Changes proposed in this PR:
- Add Infeasible Irreducible System (IIS) Tool to contrib.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

michaelbynum commented on pyomo/contrib/iis/iis.py:
```diff
@@ -0,0 +1,136 @@
+    if solver is None:
+        if len(available_solvers) == 0:
+            raise RuntimeError("Could not find a solver to use")
+        solver = _available_solvers[0]
```
is _available_solvers defined? It looks like there is no underscore above.

michaelbynum commented on pyomo/contrib/iis/iis.py:
```diff
@@ -0,0 +1,136 @@
+        if len(available_solvers) == 0:
+            raise RuntimeError("Could not find a solver to use")
+        solver = _available_solvers[0]
+        print(f"Using solver {solver}")
```
I would replace all print statements with logging statements so they can be turned off if needed.

michaelbynum commented on pyomo/contrib/iis/tests/test_iis.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class TestIIS(unittest.TestCase):
+    @unittest.skipUnless(
```
The tests look great. I would just recommend adding one more test for the case where the solver is not specified:

```
  @unittest.skipUnless(
         (pyo.SolverFactory("cplex_persistent").available(exception_flag=False) or
          pyo.SolverFactory("gurobi_persistent").available(exception_flag=False) or
          pyo.SolverFactory("xpress_persistent").available(exception_flag=False)),
         "Persistent solver not available",
     )
     def test_write_iis_any_solver(self):
         _test_iis(None)
```

I think this should work without changing any of the other test code???

michaelbynum reviewed (changes requested):
This is awesome! Thanks, @bknueven!

jsiirola commented on pyomo/contrib/iis/iis.py:
```diff
@@ -0,0 +1,136 @@
+
+
+def _remove_suffix(string, suffix):
+    if suffix in string:
```
This should be `if string.ends_with(suffix):`

jsiirola commented on pyomo/contrib/iis/iis.py:
```diff
@@ -0,0 +1,136 @@
+    "cplex_persistent": CplexConflict,
+    "gurobi_persistent": GurobiIIS,
+    "xpress_persistent": XpressIIS,
+}
```
@michaelbynum, How challenging would it be to add support for the APPSI interfaces as well?

michaelbynum replied:
Not hard for Gurobi. I still have not refactored Appsi's interface to cplex to use the python bindings.

jsiirola reviewed (commented):
Thanks, @bknueven! Some questions / requests, but overall, this is great!

bknueven commented:
Thanks for the thorough review @michaelbynum and @jsiirola. I believe I have addressed all of your comments.

I noticed incidentally that this does not quite work for Cplex 12.10 (and probably not the whole 12.x series). I may have a go at correcting that in the next few days.

I've tested locally the oldest release of Xpress available on pypi, and as for Gurobi I think its IIS API has been the same since at least version 5.6.

michaelbynum reviewed (approved):
I think it is okay if we do not support this functionality for older versions of solvers.
